diff --git a/libindicator/indicator-object.h b/libindicator/indicator-object.h
index 0c2d5e7..f8a6b89 100644
--- a/libindicator/indicator-object.h
+++ b/libindicator/indicator-object.h
@@ -36,11 +36,7 @@ typedef enum
   INDICATOR_OBJECT_SCROLL_RIGHT
 } IndicatorScrollDirection;

-#define INDICATOR_OBJECT_TYPE            (indicator_object_get_type ())
-#define INDICATOR_OBJECT(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), INDICATOR_OBJECT_TYPE, IndicatorObject))
-#define INDICATOR_OBJECT_CLASS(klass)    (G_TYPE_CHECK_CLASS_CAST ((klass), INDICATOR_OBJECT_TYPE, IndicatorObjectClass))
-#define INDICATOR_IS_OBJECT(obj)         (G_TYPE_CHECK_INSTANCE_TYPE ((obj), INDICATOR_OBJECT_TYPE))
-#define INDICATOR_IS_OBJECT_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE ((klass), INDICATOR_OBJECT_TYPE))
-#define INDICATOR_OBJECT_GET_CLASS(obj)  (G_TYPE_INSTANCE_GET_CLASS ((obj), INDICATOR_OBJECT_TYPE, IndicatorObjectClass))
+#define INDICATOR_TYPE_OBJECT (indicator_object_get_type())
+G_DECLARE_DERIVABLE_TYPE (IndicatorObject, indicator_object, INDICATOR, OBJECT, GObject)

 #define INDICATOR_OBJECT_SIGNAL_ENTRY_ADDED       "entry-added"

diff --git a/libindicator/indicator-object.c b/libindicator/indicator-object.c
index 94a5e7..d8a6b89 100644
--- a/libindicator/indicator-object.c
+++ b/libindicator/indicator-object.c
@@ -77,7 +77,7 @@ struct _IndicatorObjectPrivate {
 	GStrv environments;
 };

-#define INDICATOR_OBJECT_GET_PRIVATE(o) (INDICATOR_OBJECT(o)->priv)
+#define INDICATOR_OBJECT_GET_PRIVATE(o) indicator_object_get_instance_private(INDICATOR_OBJECT(o))

 /* Signals Stuff */
 enum {
@@ -120,7 +120,7 @@ static void indicator_object_dispose    (GObject *object);
 static void indicator_object_finalize   (GObject *object);
 static void set_property (GObject*, guint prop_id, const GValue*, GParamSpec* );

-G_DEFINE_TYPE (IndicatorObject, indicator_object, G_TYPE_OBJECT);
+G_DEFINE_TYPE_WITH_PRIVATE (IndicatorObject, indicator_object, G_TYPE_OBJECT)

 /* Setup the class and put the functions into the
    class structure */
@@ -305,7 +305,7 @@ static void
 indicator_object_init (IndicatorObject *self)
 {
-       IndicatorObjectPrivate * priv = G_TYPE_INSTANCE_GET_PRIVATE (self, INDICATOR_OBJECT_TYPE, IndicatorObjectPrivate);
+       IndicatorObjectPrivate * priv = indicator_object_get_instance_private (self);

        priv->module = NULL;

@@ -319,7 +319,6 @@ indicator_object_init (IndicatorObject *self)
        priv->entry_privates = g_hash_table_new_full (g_direct_hash, g_direct_equal, NULL, g_free);

        priv->environments = NULL;
-       self->priv = priv;
